src/XFile.cpp

```diff
@@ -11,256 +11,258 @@ namespace fs = std::filesystem;
-std::string XFile::AppendToFilename(const std::string& filename, const std::string& valueToAppend)
-{
-	fs::path newPath(filename);
+	bool HasRootComponent(const std::string& pathStr)
+	{
+		fs::path path(pathStr);
+		return path.has_root_name() || path.has_root_directory();
+	}
 
-	fs::path newFilename = newPath.filename().replace_extension("");
-	newFilename += valueToAppend + newPath.extension().string();
+	std::string MakeAbsolute(const std::string& pathStr, const std::string& relativeRootDirStr)
+	{
+		return HasRootComponent(pathStr) ? pathStr : XFile::Append(relativeRootDirStr, pathStr);
+	}
+
+	std::string Append(const std::string& path1, const std::string& relativePath2)
+	{
+		fs::path nestedPath(relativePath2);
+		if (nestedPath.has_root_name() || nestedPath.has_root_directory()) {
+			throw std::runtime_error("Cannot append paths with root component: " + relativePath2);
+		}
+		return (fs::path(path1) / nestedPath).generic_string();
+	}
 
-	newPath.replace_filename(newFilename);
+	std::string AppendToFilename(const std::string& filename, const std::string& valueToAppend)
+	{
+		fs::path newPath(filename);
 
-	return newPath.string();
-}
+		fs::path newFilename = newPath.filename().replace_extension("");
+		newFilename += valueToAppend + newPath.extension().string();
 
-std::vector<std::string> XFile::Dir(const std::string& directory)
-{
-	// Creating a path with an empty string will prevent the directory_iterator from finding files in the current relative path.
-	auto pathStr = directory.length() > 0 ? directory : "./";
+		newPath.replace_filename(newFilename);
 
-	std::vector<std::string> filenames;
-	for (const auto& entry : fs::directory_iterator(pathStr)) {
-		filenames.push_back(GetFilename(entry.path().generic_string()));
+		return newPath.string();
 	}
 
-	return filenames;
-}
+	std::vector<std::string> Dir(const std::string& directory)
+	{
+		// Creating a path with an empty string will prevent the directory_iterator from finding files in the current relative path.
+		auto pathStr = directory.length() > 0 ? directory : "./";
 
-template <typename SelectFunction>
-std::vector<std::string> Dir(const std::string& directory, SelectFunction selectFunction)
-{
-	// Creating a path with an empty string will prevent the directory_iterator from finding files in the current relative path.
-	auto pathStr = directory.length() > 0 ? directory : "./";
-
-	std::vector<std::string> filenames;
-	for (const auto& entry : fs::directory_iterator(pathStr)) {
-		auto pathString = entry.path().generic_string();
-		if (selectFunction(pathString)) {
-			filenames.push_back(XFile::GetFilename(pathString));
+		std::vector<std::string> filenames;
+		for (const auto& entry : fs::directory_iterator(pathStr)) {
+			filenames.push_back(GetFilename(entry.path().generic_string()));
 		}
-	}
 
-	return filenames;
-}
+		return filenames;
+	}
 
-std::vector<std::string> XFile::Dir(const std::string& directory, const std::regex& filenameRegex)
-{
-	return ::Dir(
```

Comment:
@DanRStevens I had to remove the global namespace specifier from the Dir function calls in XFile when I moved the namespace syntax from displaying on each function to a block style namespace call.

I wasn't sure while the global namespace specifier was there in the first place. Wanted to call this out considering how much noise there is in the commits.

Reply:
It's there because there are multiple methods with the name `Dir`, some in the `XFile` namespace, and one in the global namespace (defined privately in XFile.cpp). Since the calling code is within a method in the `XFile` namespace, it seems to be resolving the name to the overloads defined in the `XFile` namespace, when the global template method was intended.